File: leetcode/8.StringToInteger(atoi).go
package leetcode

import (
	"math"
	"strconv"
)

func myAtoi(s string) int {
	result := 0
	sign := 1
	found := false

	for _, char := range s {
		word := string(char)

		isWhiteSpace := word == " "
		hasWhiteSpaceWithinNumbers := isWhiteSpace && found
		isPlusSign := word == "+"
		isMinusSign := word == "-"

		if hasWhiteSpaceWithinNumbers {
			break
		}

		if isWhiteSpace {
			continue
		}

		if _, err := strconv.Atoi(word); err == nil {
			found = true
			result = addNumber(result, word)
			continue
		}

		if isPlusSign && !found {
			found = true
			continue
		}

		if isMinusSign && !found {
			found = true
			sign *= -1
			continue
		}

		break
	}

	result *= sign

	max := math.Max(float64(-int(math.Pow(float64(2), float64(31)))), float64(result))

	return int(math.Min(max, float64(int(math.Pow(2, 31))-1)))
}

func addNumber(n int, val string) int {
	num, _ := strconv.Atoi(val)

	return n*10 + num
}

File: leetcode/8.StringToInteger(atoi).ts
function myAtoi(s: string) {
  let result = 0
  let sign = 1
  let found = false

  const addNumber = (n: number, val: string) => n * 10 + Number(val)

  for (let i = 0; i < s.length; i++) {
    const char = s[i]
    const isWhiteSpace = char === ' '
    const hasWhiteSpaceWithinNumbers = isWhiteSpace && found
    const isNumeric = char >= '0' && char <= '9'
    const isPlusSign = char === '+'
    const isMinusSign = char === '-'

    if (hasWhiteSpaceWithinNumbers) break
    if (isWhiteSpace) continue

    if (isNumeric) {
      found = true
      result = addNumber(result, char)
      continue
    }

    if (isPlusSign && !found) {
      found = true
      continue
    }

    if (isMinusSign && !found) {
      found = true
      sign *= -1
      continue
    }

    break
  }

  result *= sign
  return Math.min(Math.max(-(2 ** 31), result), 2 ** 31 - 1)
}
console.log(myAtoi('   +0 123'))

function myAtoiByParseInt(s: string) {
  const trimmed = s.trim()
  if (trimmed === '') return 0

  const int = parseInt(trimmed, 10)
  if (isNaN(int)) return 0

  const min = Math.pow(-2, 31)
  if (int < min) return min

  const max = Math.pow(2, 31) - 1
  if (int > max) return max

  return int
}
console.log(myAtoiByParseInt('     4193 with words    '))
